### ISSUE TITLE
Allow `matchesSemantics` to only match certain properties

### ISSUE THREAD

**Issue by bryanoltman** [flutter/flutter#107859]
## Use case

When writing a widget test that verifies the semantic properties of an element, I'm forced to specify every semantics trait the corresponding SemanticsNode has. For example, if I simply want to verify that the element found by `find.text('myButton')` is seen as a button in the semantics tree, I have to write:

```dart
expect(
  tester.getSemantics(find.text('myButton')),
  matchesSemantics(
    label: 'myButton',
    isButton: true,
    hasTapAction: true,
    textDirection: TextDirection.ltr,
    isEnabled: true,
    isFocusable: true,
    hasEnabledState: true,
  ),
);
```

This is inconvenient for the developer, makes the purpose of the test less clear, and makes the test more brittle, as changes to other semantic values that are irrelevant to this test will cause this test to start failing.

## Proposal

Introduce a matcher that only checks the values provided to the matcher. For example, if we were to call such a matcher `hasSemantics`, the test above could be written as:

```dart
expect(
  tester.getSemantics(find.text('myButton')),
  hasSemantics(isButton: true),
);
```

**Comment by franckbleyer1** [comment]
Awww yeahhh ;-) .

**Comment by jonahwilliams** [comment]
I think we could probably change the API such that each default value is actually nullable, and then ignore nullable properties in the comparison. That should be backwards compatible

**Comment by bryanoltman** [comment]
@jonahwilliams that would work too! :)

**Comment by goderbauer** [comment]
I support the idea of being able to match partial semantics. That way you can only encode the things in your test that actually matter without codifying the properties that the test doesn't actually care about.

> I think we could probably change the API such that each default value is actually nullable, and then ignore nullable properties in the comparison. That should be backwards compatible

Sadly, I think this wouldn't be backwards compatible. Due to the lint `avoid_redundant_argument_values` someone may actually assert that something is NOT a button by just writing `hasSemantics()` since `isButton` defaults to false. This currently makes the right assertion. If we now change the default value of `isButton` to null, the test is no longer making this expected assertion and the semantics of the test have changed for the worse.

**Comment by pdblasi-google** [comment]
@bryanoltman @goderbauer 

I ended up having to go with a name other than `hasSemantics` (I chose `containsSemantics`) because there was a naming conflicts with other global functions in some tests in the repo.
